Updating nodes page per style suggestion (#23767)
kubernetes/website: content/en/docs/concepts/architecture/nodes.md

<!-- overview -->

> Kubernetes runs your workload by placing containers into pods to run on _nodes_.

Reviewer: My reading of this sentence (as formatted) is that Node is being introduced as a concept.
Perhaps the italics is not needed?
When I read the word, nodes (lowercase), I think of a machine or VM.

Reviewer: I don't know if we italicize introductory terms without also capitalizing, but to move forward let's leave this.

Reviewer 2: > I don't know if we italicize introductory terms without also capitalizing, but to move forward let's leave this.

Sometimes we do.

## Management

> There are two main ways to have nodes added to the {{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}:

Reviewer: When I read this sentence, it seems as if the nodes are physically added to the API server.
Instead:

There are two ways to add nodes to the API server:
The kubelet running on a node registers with the control plane
A user manually adds a `Node` using [add how this is added]

Remove "object" after resource name:
After you create a `Node` or the kubelet on the node self-registers, ...
the new `Node` is valid.

Reviewer: OK.

Reviewer 2: How about:
```suggestion
Before the control plane can [schedule](/docs/concepts/scheduling-eviction/) any pods to run on
a node, the node needs to be registered with the cluster. There are two common ways to
register nodes with the control plane:
1.  The kubelet, running on a node, [registers itself](#self-registration-of-nodes) with
    the cluster's
    {{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}.
1.  A cluster administrator [manually](#manual-node-administration) adds a `Node` using a tool such as `kubectl`.
```

Now ideally https://k8s.io/docs/concepts/scheduling-eviction/ should include the glossary definition for “scheduling”, but I'd rather link there now and let the smart reader click through to the first page in that section if they wanted to learn more.

Reviewer 2: @bergerhoffer when you're adding it to the API server, it's a Node (an object with `kind: Node`).

}
```

Kubernetes creates a `Node` object internally (the representation). Kubernetes checks
that a kubelet has registered to the API server that matches the `metadata.name`
field of the `Node`. If the node is healthy (if all necessary services are running),

> it is eligible to run a pod. Otherwise, that node is ignored for any cluster activity

Reviewer: nit: Question about using lowercase for Pod.

Reviewer 2: Either `Pod` or pod is fine. I would pick `Pod`.

until it becomes healthy.

{{< note >}}

> Kubernetes keeps the object for the invalid node and continues checking to see whether

Reviewer: nit: what object is kept?

Reviewer: nit: Kubernetes keeps the `Node` object for the invalid node and ...

it becomes healthy.

You, or a {{< glossary_tooltip term_id="controller" text="controller">}}, must explicitly

> delete the `Node` object to stop that health checking.

Reviewer: I would not use both monospace and object with the resource, Node.

Reviewer: If it helps clarify the concept, okay to update.

    No-op if `register-node` is false.
  - `--node-ip` - IP address of the node.
  - `--node-labels` - {{< glossary_tooltip text="Labels" term_id="label" >}} to add when registering the node in the cluster (see label restrictions enforced by the [`NodeRestriction` admission plugin](/docs/reference/access-authn-authz/admission-controllers/#noderestriction)).
  - `--node-status-update-frequency` - Specifies how often kubelet posts node status to master.

> When the [node authorization mode](/docs/reference/access-authn-authz/node/) and

Reviewer: - I would capitalize Node (line 86)
- Line 88, this is an odd sentence, "... kubelets are only authorized ..."

Reviewer 2: ```suggestion
When the [`Node` authorization mode](/docs/reference/access-authn-authz/node/) and
```
or
```markdown
Once you enable the [Node](/docs/reference/access-authn-authz/node/) authorization mode and
```
_plus further changes to match_.

It's “Node” when you configure it in the API server: `--authorization-mode=Node`

[`NodeRestriction` admission plugin](/docs/reference/access-authn-authz/admission-controllers/#noderestriction) are enabled,
kubelets are only authorized to create/modify their own `Node` resource.

> ### Manual node administration

Reviewer: This section refers to the Node API object. Why should the heading lowercase node?

Reviewer 2: The section describes the API changes you might make before rebooting the node or doing other maintenance.
(and that you would later undo).
So, although you definitely do interact with Node objects here, the overall context for the task is “node administration”.

Reviewer: Oh. Could the heading read, "Manual [virtual] machine administration".

You can create and modify `Node` objects using
{{< glossary_tooltip text="kubectl" term_id="kubectl" >}}.

When you want to create `Node` objects manually, set the kubelet flag `--register-node=false`.

You can modify `Node` objects regardless of the setting of `--register-node`.
For example, you can set labels on an existing `Node`, or mark it unschedulable.

> You can use labels on nodes in conjunction with node selectors on pods to control

Reviewer: Not sure about this modification.

Reviewer 2: I agree; the original text is referring to the objects, so I'd write:
```suggestion
You can use labels on `Nodes` in conjunction with node selectors on `Pods` to influence
```

scheduling. For example, you can constrain a pod to only be eligible to run on
a subset of the available nodes.

Marking a node as unschedulable prevents the scheduler from placing new pods onto
that node, but does not affect existing pods on the node. This is useful as a
preparatory step before a node reboot or other maintenance.

To mark a node unschedulable, run:

```shell
kubectl cordon $NODENAME
```

{{< note >}}
Pods that are part of a {{< glossary_tooltip term_id="daemonset" >}} tolerate

> being run on an unschedulable node. Daemon sets typically provide node-local services

Reviewer: ? "Daemon sets" == sets of daemons?

Reviewer 2: “DaemonSets” might work better, I reckon.

Reviewer: Yes, @bergerhoffer , Would you revert this change to `DaemonSets`?

that should run on the node even if it is being drained of workload applications.
{{< /note >}}

## Node status

> A node's status contains the following information:

Reviewer: ?

Reviewer 2: Good spot!

Reviewer: @bergerhoffer , Are you open to changing this?
See, https://deploy-preview-23767--kubernetes-io-master-staging.netlify.app/docs/reference/generated/kubernetes-api/v1.19/#nodestatus-v1-core

Possible edit:
The `Node` resource contains a `NodeStatus` object with the following information:

Reviewer 2: My suggestion```suggestion
Each `Node` resource contains a `.status` field that includes the following information:
```
Optionally, also make `.status` a hyperlink to NodeStatus in the API reference.

The usage of these fields varies depending on your cloud provider or bare metal configuration.

> * `HostName`: The host name as reported by the node's kernel. Can be overridden via the kubelet `--hostname-override` parameter.

Reviewer: nit: style, s/via/by/

{{< /table >}}

{{< note >}}
If you use command-line tools to print details of a cordoned node, the condition includes

> `SchedulingDisabled`. `SchedulingDisabled` is not a condition in the Kubernetes API; instead,

Reviewer: Here "condition in the Kubernetes API" refers to the `Condition` object in the Kubernetes API?

]
```

> If the status of the `Ready` condition remains `Unknown` or `False` for longer than the `pod-eviction-timeout` (an argument passed to the {{< glossary_tooltip text="kube-controller-manager" term_id="kube-controller-manager" >}}), all the pods on the node are scheduled for deletion by the node controller. The default eviction timeout duration is **five minutes**. In some cases when the node is unreachable, the API server is unable to communicate with the kubelet on the node. The decision to delete the pods cannot be communicated to the kubelet until communication with the API server is re-established. In the meantime, the pods that are scheduled for deletion may continue to run on the partitioned node.

Reviewer: This may be the `status` field?

Reviewer 2: I think so, yes.

Reviewer: This can be confusing, because there is a status field and the "status of a condition" also generically makes sense.
This paragraph follows the JSON definition of a `Condition` object and I think tries to explain the "condition" using the fields in the previous definition.
Is this any better:
If the `Condition` is in a `Ready` state but the `Condition.status` remains `Unknown` or `False` for longer than

The node controller does not force delete pods until it is confirmed that they have stopped
running in the cluster. You can see the pods that might be running on an unreachable node as
being in the `Terminating` or `Unknown` state. In cases where Kubernetes cannot deduce from the
underlying infrastructure if a node has permanently left a cluster, the cluster administrator
may need to delete the `Node` object by hand.  Deleting the `Node` object from Kubernetes causes
all the `Pod` objects running on the node to be deleted from the API server, and frees up their
names.

The node lifecycle controller automatically creates
[taints](/docs/concepts/scheduling-eviction/taint-and-toleration/) that represent conditions.

> The scheduler takes the node's taints into consideration when assigning a pod to a node.

Reviewer: I question, "node's taints"

controller deletes the node from its list of nodes.

The third is monitoring the nodes' health. The node controller is
responsible for updating the `NodeReady` condition of `NodeStatus` to
`ConditionUnknown` when a node becomes unreachable (i.e. the node controller stops
receiving heartbeats for some reason, for example due to the node being down), and then later evicting
all the pods from the node (using graceful termination) if the node continues
to be unreachable. (The default timeouts are 40s to start reporting

> `ConditionUnknown` and 5m after that to start evicting pods.) The node controller

Reviewer: This could use more cleanup.
Okay to save for another PR.
nit: style, `ConditionUnknown` when a node becomes unreachable (for example, the node controller stops
receiving heartbeats because the node has shutdown),

checks the state of each node every `--node-monitor-period` seconds.

#### Heartbeats

Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.

There are two forms of heartbeats: updates of `NodeStatus` and the
[`Lease` object](/docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io).

> Each node has an associated `Lease` object in the `kube-node-lease`

Reviewer: What do you think about changing to:
Each `Node` object has an associate `Lease` object ...

Reviewer 2: ```suggestion
Each `Node` has an associated `Lease`, in the `kube-node-lease`
```

If you have enabled the `TopologyManager`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/), then
the kubelet can use topology hints when making resource assignment decisions.

> See [Control Topology Management Policies on a node](/docs/tasks/administer-cluster/topology-manager/)

Reviewer: What do you think about reverting this change?
Since this is anchor text and all words are capitalized, leave Node capitalized?

for more information.

## {{% heading "whatsnext" %}}

* Learn about the [components](/docs/concepts/overview/components/#node-components) that make up a node.

> * Read the [API definition for `Node`](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#node-v1-core).

Reviewer: nit: Read about the `[Node` API](...) ?

Reviewer 2: I'm OK with the change as drafted here.
